crates/red_knot_python_semantic/src/types.rs

```diff
@@ -6133,12 +6144,18 @@ impl<'db> BoolError<'db> {
     }
 }
 
+#[derive(Debug)]
+enum DunderNewCallError<'db> {
+    CallError(CallError<'db>),
+    PossiblyUnbound(Option<CallError<'db>>),
+}
```

Comment:
am I correct in saying that this represents the fact that there are three possible ways a `__new__` call could generate a diagnostic?

1. It could be called with incorrect arguments
2. It could be possibly unbound
3. It could be called with incorrect arguments AND be possibly unbound

I think the enum structure is fine, but you could possibly add a doc-comment to the enum as a whole